feat: Add variables input validation and permissions (no-changelog)

alexgrozav opened the pull request:
<img width="1316" alt="image" src="https://user-images.githubusercontent.com/6179477/230121767-3ebef744-366d-4181-9877-de62b364cafa.png">
<img width="1255" alt="image" src="https://user-images.githubusercontent.com/6179477/230122404-a0051934-9bfc-4c28-97d4-d37acd742f2d.png">
<img width="1249" alt="image" src="https://user-images.githubusercontent.com/6179477/230122029-c87510b7-3215-429e-b7cf-71e8035c886a.png">

alexgrozav committed:
feat: add input validation

alexgrozav committed:
feat: handle variables view for non-instance-owner users

alexgrozav committed:
test: update variables tests

alexgrozav committed:
fix: fix data-testid pattern

alexgrozav committed:
feat: improve overflow styles

alexgrozav committed:
test: fix variables row snapshot

alexgrozav committed:
feat: update sorting to take newly created variables into account

alexgrozav committed:
fix: fix list layout overflow

alexgrozav committed:
fix: fix adding variables on page other than 1. fix validation

alexgrozav committed:
feat: add docs link

ivov commented on packages/editor-ui/src/views/VariablesView.vue:
```diff
@@ -155,33 +197,37 @@ function displayName(resource: EnvironmentVariable) {
 	>
 		<template #add-button>
-			<n8n-tooltip placement="top" :disabled="isFeatureEnabled">
+			<n8n-tooltip placement="top" :disabled="canCreateVariables">
```
I'd have expected `!canCreateVariables`?

ivov reviewed (commented):
No context on this feature so please disregard any comments if not relevant:

1. Sorting of workflow cards does not seem to work anymore. Did not test master.
2. Searching of workflow cards does not seem to work anymore. Did not test master.
3. ~~Value can be empty and dispayed as empty. This is intended and stands for undefined? (If so, I find this confusing.) Also null can only exist as a string?~~ Edit: All values are strings. Expanded in comment 8.
4. If I pick `All` in the pager and then add a variable, `Page Size` becomes the actual number of entries instead of remaining as `All`.

ivov commented:
Many of Val's methods do not work with $vars and should probably be hidden?

<img width="906" alt="image" src="https://user-images.githubusercontent.com/44588767/231401103-1a868798-40be-4fee-af8a-7d16064988da.png">

ivov commented:
6. Overflow with value having max chars:

<img width="1017" alt="image" src="https://user-images.githubusercontent.com/44588767/231402285-124391e9-a52d-4a8f-9c6f-4e628f6d92e0.png">

ivov commented:
7. An n8n var name may begin with a digit, but this is illegal JS:

<img width="1034" alt="image" src="https://user-images.githubusercontent.com/44588767/231403132-7c32ecde-f833-4720-b0e8-7e0c3bb5fb80.png">

ivov commented:
8. UX wise, the fact that all env var values resolve to strings is bound to send users debugging often:

<img width="1034" alt="image" src="https://user-images.githubusercontent.com/44588767/231406073-321d92a8-fe22-425d-880a-aae4d3837bb4.png">

alexgrozav committed:
fix: fix default displayName function for resource-list-layout

alexgrozav commented on packages/editor-ui/src/components/VariablesRow.vue:
```diff
@@ -53,51 +88,80 @@ async function onSave() {
-				v-model="modelValue.key"
-				ref="keyInputRef"
-			/>
+		<td class="variables-key-column">
```
The most common use case is to make non-obfuscated class names available in another component. 

For example, if we want to style the VariablesRow in a specific usage context, we'll be able to modify the classes we expose. Here, the choice was made so that we're able to modify the classes for both the `thead tr th` and the `tbody tr td` inside of the VariablesView.

alexgrozav commented on packages/editor-ui/src/views/VariablesView.vue:
```diff
@@ -155,33 +197,37 @@ function displayName(resource: EnvironmentVariable) {
 				</div>
 				<template #content>
-					{{ i18n.baseText('variables.add.unavailable') }}
+					<span v-if="!isFeatureEnabled">{{ i18n.baseText('variables.add.unavailable') }}</span>
```
The tooltip only shows when the user can't create variables, in the following 2 cases:
A. The user cannot create variables because the feature is not enabled -> show sales message
B. The user cannot create variables because the user does not have permissions -> show that instance owner can create variables

alexgrozav commented:
> 5. Many of Val's methods do not support `$vars` and so should probably be hidden?
> 
> <img alt="image" width="906" src="https://user-images.githubusercontent.com/44588767/231401103-1a868798-40be-4fee-af8a-7d16064988da.png">

Refactored how `$vars` is added to expressions.

> 6. Overflow with value having max chars:
> 
> <img alt="image" width="1017" src="https://user-images.githubusercontent.com/44588767/231402285-124391e9-a52d-4a8f-9c6f-4e628f6d92e0.png">

Added overflow handling.



> 7. An n8n var name may begin with a digit, but this is illegal JS:
> 
> <img alt="image" width="1034" src="https://user-images.githubusercontent.com/44588767/231403132-7c32ecde-f833-4720-b0e8-7e0c3bb5fb80.png">

You could access it as `$vars['123']`. Need to discuss with Romain.



> 8. UX wise, the fact that all env var values resolve to strings is bound to send users debugging often:
> 
> <img alt="image" width="1034" src="https://user-images.githubusercontent.com/44588767/231406073-321d92a8-fe22-425d-880a-aae4d3837bb4.png">

I think variable types are going to be handled at a later stage? Need to discuss with Romain.

alexgrozav committed:
feat: improve vars expressions ux, cm-tooltip

alexgrozav committed:
test: fix datatable test

ivov reviewed (commented):
Thanks for addressing everything!

Some more findings, but please disregard as you need:

[4] I can still reproduce this. From what I can tell, the only valid options should be `10`, `25`, `50`, `100` and `All`
[5] As a quick win, we can enable all string completions for `$vars.someVar`. This would be handy for the user and help reinforce these are all strings.
[7] It is possible to access it via `$vars[123]` or `$vars['123']` but both autocompletion and the usage syntax column  are still proposing `$vars.123`, which we know will throw.
[9] If [this](https://n8nio.slack.com/archives/C04NH65H7CM/p1681306380155679?thread_ts=1681305040.769179&cid=C04NH65H7CM) is the latest, then validation should require starting with a letter. This is not enforced right now.
[10] If [this](https://n8nio.slack.com/archives/C04NH65H7CM/p1681306380155679?thread_ts=1681305040.769179&cid=C04NH65H7CM) is the latest, then the doc snippet should be `Variable set on this n8n instance. All variables evaluate to strings.` This is not the copy right now.
[11] Odd that we update usage syntax [even when the key name fails validation](https://user-images.githubusercontent.com/44588767/231693681-7bbdcb29-b4b9-4398-a425-17d298edbfc7.png).
[12] Removing license for an instance that had vars added shows [both upsell message and existing vars](https://user-images.githubusercontent.com/44588767/231693815-9683506a-2cfa-4a12-ab64-c54efcf86709.png). I'd expect the table to be hidden. Doublechecking if intended. 
[13] Removing license for an instance that had vars still autocompletes `$vars` and still allows for normal usage, i.e. it still resolves values. I'd expect all this to be disabled. Doublechecking if intended.
[14] Deleting a variable (and arguably editing it as well) should mention in bold that this can break workflows, same as we do with credential deletion or unsharing.
[15] Upsell message: `Variables are immutable and cannot be modified within your workflows` I find this copy misleading - variables are easily editable and deletable, potentially breaking your workflow. Immutability to me would be a guarantee that any vars I use cannot change under my feet.
[16] If I'm new to the feature and have a matching license (i.e. hence not seeing the upsell message, potentially maybe never even saw it at all), then I'd appreciate some context at `/variables`. A short explanatory note about what this feature is, just like in the upsell message? Or a link to the docs? Instead of displaying the table and assuming this is self-explanatory.

~~Have not had time to check the code yet.~~ Edit: Done

ivov reviewed (approved):
Code looks good! Approving regardless of the above in case you need to move forward with this.

alexgrozav committed:
Merge branch 'pay-230-p0-variables' into pay-235-variables-validation

alexgrozav committed:
feat: add MATCH_REGEX validation rule

alexgrozav committed:
fix: overhaul how datatable pagination selector works

alexgrozav committed:
feat: update  completer description

alexgrozav committed:
fix: conditionally update usage syntax based on key validation

alexgrozav commented:
>[4] I can still reproduce this. From what I can tell, the only valid options should be 10, 25, 50, 100 and All
I see what you mean now. Rehauled pagination handler to use `*` for `All`. Also, scrolling the layout to the top when adding new vars.

>[5] As a quick win, we can enable all string completions for $vars.someVar. This would be handy for the user and help reinforce these are all strings.
Just checked. Seems to have string methods already if you write `$vars.someVar.`. Am I misunderstanding this?

>[7] It is possible to access it via $vars[123] or $vars['123'] but both autocompletion and the usage syntax column are still proposing $vars.123, which we know will throw.
Yup. Added validator.

> [9] If [this](https://n8nio.slack.com/archives/C04NH65H7CM/p1681306380155679?thread_ts=1681305040.769179&cid=C04NH65H7CM) is the latest, then validation should require starting with a letter. This is not enforced right now.
Wasn't latest. Updated.

> [10] If [this](https://n8nio.slack.com/archives/C04NH65H7CM/p1681306380155679?thread_ts=1681305040.769179&cid=C04NH65H7CM) is the latest, then the doc snippet should be Variable set on this n8n instance. All variables evaluate to strings. This is not the copy right now.
Wasn't latest. Updated.

> [11] Odd that we update usage syntax [even when the key name fails validation](https://user-images.githubusercontent.com/44588767/231693681-7bbdcb29-b4b9-4398-a425-17d298edbfc7.png).
Fair point. Updated how usage syntax is computed.


>[12] Removing license for an instance that had vars added shows [both upsell message and existing vars](https://user-images.githubusercontent.com/44588767/231693815-9683506a-2cfa-4a12-ab64-c54efcf86709.png). I'd expect the table to be hidden. Doublechecking if intended.
This is intended.

> [13] Removing license for an instance that had vars still autocompletes $vars and still allows for normal usage, i.e. it still resolves values. I'd expect all this to be disabled. Doublechecking if intended.
This is intended. Only adding/editing/removing is disabled. Usage is allowed.

>[14] Deleting a variable (and arguably editing it as well) should mention in bold that this can break workflows, same as we do with credential deletion or unsharing.
Needs discussion with Romain and Giulio.

> [15] Upsell message: Variables are immutable and cannot be modified within your workflows I find this copy misleading - variables are easily editable and deletable, potentially breaking your workflow. Immutability to me would be a guarantee that any vars I use cannot change under my feet.
Needs discussion with Romain and Giulio.


[16] If I'm new to the feature and have a matching license (i.e. hence not seeing the upsell message, potentially maybe never even saw it at all), then I'd appreciate some context at /variables. A short explanatory note about what this feature is, just like in the upsell message? Or a link to the docs? Instead of displaying the table and assuming this is self-explanatory.

We do have an empty state, but probably copy sohuld be updated. Needs discussion with Romain and Giulio.
<img width="1005" alt="image" src="https://user-images.githubusercontent.com/6179477/231727285-5e564a03-efe5-4ed6-9e11-094e0d4dae95.png">

alexgrozav committed:
test: update datatable snapshot

alexgrozav committed:
fix: fix variables-row button margins

alexgrozav committed:
fix: fix pagination overflow

ivov reviewed (approved):
All working, thanks for addressing everything!

[5] I should have specified - the string method completions are missing for `$vars.someVar.` only in the code node editor. This can be added back later though.
[16] The empty state will likely only be visible for the owner at the beginning, i.e. a member will likely never see this empty state, so I was thinking of an infotip with a 1-line explainer like we have elsewhere. Very minor though.
